From 7c95f51ad4733a5a9ec0075d63420091ae2c5f87 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:04:21 -0400
Subject: [powerpc] pci: fix PCI-e check link issue

Message-id: <1410545655-205645-33-git-send-email-dzickus@redhat.com>
Patchwork-id: 94080
O-Subject: [RHEL7 PATCH 032/626] powerpc/pci: fix PCI-e check link issue
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 36f684940017b7a5d9039861189203d64d2f8861
Author: Yuanquan Chen <Yuanquan.Chen@freescale.com>
Date:   Fri May 17 15:35:29 2013 +0800

    powerpc/pci: fix PCI-e check link issue

    For Freescale powerpc platform, the PCI-e bus number uses the reassign mode
    by default. It means the second PCI-e controller's hose->first_busno is the
    first controller's last bus number adding 1. For some hotpluged device(or
    controlled by FPGA), the device is linked to PCI-e slot at linux runtime.
    It needs rescan for the system to add it and driver it to work. It successes
    to rescan the device linked to the first PCI-e controller's slot, but fails to
    rescan the device linked to the second PCI-e controller's slot. The cause is
    that the bus->number is reset to 0, which isn't equal to the hose->first_busno
    for the second controller checking PCI-e link. So it doesn't really check the
    PCI-e link status, the link status is always no_link. The device won't be
    really rescaned. Reset the bus->number to hose->first_busno in the function
    fsl_pcie_check_link(), it will do the real checking PCI-e link status for the
    second controller, the device will be rescaned.

    Signed-off-by: Yuanquan Chen <Yuanquan.Chen@freescale.com>
    Tested-by: Rojhalat Ibrahim <imr@rtschenk.de>
    Signed-off-by: Scott Wood <scottwood@freescale.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/sysdev/fsl_pci.c b/arch/powerpc/sysdev/fsl_pci.c
index 5402a1d..89d99e6 100644
--- a/arch/powerpc/sysdev/fsl_pci.c
+++ b/arch/powerpc/sysdev/fsl_pci.c
@@ -64,7 +64,7 @@ static int fsl_pcie_check_link(struct pci_controller *hose)
  if (hose->indirect_type & PPC_INDIRECT_TYPE_FSL_CFG_REG_LINK) {
   if (hose->ops->read == fsl_indirect_read_config) {
    struct pci_bus bus;
-   bus.number = 0;
+   bus.number = hose->first_busno;
    bus.sysdata = hose;
    bus.ops = hose->ops;
    indirect_read_config(&bus, 0, PCIE_LTSSM, 4, &val);
-- 
1.7.1